url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+     * Note that the result could also be null, it is a valid result.
+     *
+     * @return
+     */
+    public synchronized boolean hasResult()
+    {
+        // No guard because you always could get the last result for examination,
+        // even post mortem.
+
+        try
+        {
+            while(!halted) this.wait();
+            return hasResult;
+        }
+        catch (InterruptedException e)
+        {
+            // The wait could be interrupted because the
+            // eval thread ended.
+            halted = true;
+            return hasResult;
+        }
+    }
+
+    /**
+     * Stop tracing. No operation is possible after the tracer has halted.
+     * All resources held by the eval being traced will be released.
+     * Stopping a tracer is necessary in order not to clutter the memory.
+     *
+     */
+   public synchronized void terminate()
+   {
+       if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+       try
+        {
+            while(!halted) this.wait();
+            worker.interrupt();
+            worker = null;
+            halted = true;
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+        }
+   }
+
+   /**
+    * Check if the trace has been halted or not.
+    * If it is halted, nothing can be done with it anymore.
+    *
+    * @return
+    */
+   public synchronized boolean isTerminated()
+   {
+       return (worker == null);
+   }
+
+   /**
+    * Restart the evaluation of the current expression to the beginning.
+    * Ignore all intermediate results obtained so far.
+    *
+    */
+   public synchronized void reset()
+   {
+       if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+       try
+        {
+            while(!halted) this.wait();
+            stack.reset();
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+        }
+   }
+
+   /**
+    * Restart the evaluation of the current expression to the beginning.
+    * Ignore all intermediate results obtained so far.
+    *
+    */
+   public synchronized void dropFrame()
+   {
+       if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+       try
+        {
+            while(!halted) this.wait();
+            stack.dropFrame();
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+        }
+   }
+
+   /**
+    * Step backwards. Note that side effects will not be undone.
+    *
+    */
+   public synchronized void backStep()
+   {
+       if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+       try
+        {
+            while(!halted) this.wait();
+
+            final Eval2.StackFrame lFrame = stack.top();
+            if(lFrame.getDataptr() > 0) lFrame.backStep();
+            else stack.dropFrame();
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)